builtin/providers/azurerm/resource_arm_template_deployment.go

```diff
@@ -0,0 +1,212 @@
+package azurerm
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/arm/resources/resources"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceArmTemplateDeployment() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmTemplateDeploymentCreate,
+		Read:   resourceArmTemplateDeploymentRead,
+		Update: resourceArmTemplateDeploymentCreate,
+		Delete: resourceArmTemplateDeploymentDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"resource_group_name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"template_body": &schema.Schema{
+				Type:      schema.TypeString,
+				Optional:  true,
+				Computed:  true,
+				StateFunc: normalizeJson,
+			},
+
+			"parameters": &schema.Schema{
+				Type:     schema.TypeMap,
+				Optional: true,
+			},
+
+			"outputs": &schema.Schema{
+				Type:     schema.TypeMap,
+				Computed: true,
+			},
+
+			"deployment_mode": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+		},
+	}
+}
+
+func resourceArmTemplateDeploymentCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient)
+	deployClient := client.deploymentsClient
+
+	name := d.Get("name").(string)
+	resGroup := d.Get("resource_group_name").(string)
+	deploymentMode := d.Get("deployment_mode").(string)
+
+	log.Printf("[INFO] preparing arguments for Azure ARM Template Deployment creation.")
+	properties := resources.DeploymentProperties{
+		Mode: resources.DeploymentMode(deploymentMode),
+	}
+
+	if v, ok := d.GetOk("parameters"); ok {
+		params := v.(map[string]interface{})
+
+		newParams := make(map[string]interface{}, len(params))
+		for key, val := range params {
+			newParams[key] = struct {
+				Value interface{}
+			}{
+				Value: val,
+			}
+		}
+
+		properties.Parameters = &newParams
+	}
+
+	if v, ok := d.GetOk("template_body"); ok {
+		template, err := expandTemplateBody(v.(string))
+		if err != nil {
+			return err
+		}
+
+		properties.Template = &template
+	}
+
+	deployment := resources.Deployment{
+		Properties: &properties,
+	}
+	resp, err := deployClient.CreateOrUpdate(resGroup, name, deployment)
+	if err != nil {
+		return nil
+	}
+
+	d.SetId(*resp.ID)
+
+	log.Printf("[DEBUG] Waiting for Template Deployment (%s) to become available", name)
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{"Creating", "Updating", "Accepted", "Running"},
+		Target:  []string{"Succeeded"},
```

Comment:
This casing feels brittle - I think we should make this all lower and have the `RefreshFunc` return `strings.ToLower()` of the status.